utils/testrunner/TestRunner.js

```diff
@@ -156,27 +156,41 @@ class Test {
   }
 
   environment(environment) {
+    const parents = new Set();
+    for (let parent = environment; !(parent instanceof Suite); parent = parent.parentEnvironment())
+      parents.add(parent);
+    for (const env of this._environments) {
+      for (let parent = env; !(parent instanceof Suite); parent = parent.parentEnvironment()) {
+        if (parents.has(parent))
+          throw new Error(`Cannot use environments "${environment.name()}" and "${env.name()}" that share a parent environment "${parent.fullName()}" in test "${this.fullName()}"`);
+      }
+    }
+    const environmentParentSuite = environment.parentSuite();
     for (let suite = this.suite(); suite; suite = suite.parentSuite()) {
-      if (suite === environment.parentSuite()) {
+      if (suite === environmentParentSuite) {
         this._environments.push(environment);
-        return;
+        return this;
       }
     }
     throw new Error(`Cannot use environment "${environment.name()}" from suite "${environment.parentSuite().fullName()}" in unrelated test "${this.fullName()}"`);
   }
 }
 
 class Environment {
-  constructor(parentSuite, name, location) {
-    this._parentSuite = parentSuite;
+  constructor(parentEnvironment, name, location) {
+    this._parentEnvironment = parentEnvironment;
     this._name = name;
-    this._fullName = (parentSuite ? parentSuite.fullName() + ' ' + name : name).trim();
+    this._fullName = (parentEnvironment ? parentEnvironment.fullName() + ' ' + name : name).trim();
     this._location = location;
     this._hooks = [];
   }
 
+  parentEnvironment() {
+    return this._parentEnvironment;
+  }
+
   parentSuite() {
-    return this._parentSuite;
+    return (!this._parentEnvironment || this._parentEnvironment instanceof Suite) ? this._parentEnvironment : this._parentEnvironment.parentSuite();
```

Comment:
hidden O(N)! Shall we call this `computeParentSuite()`? or maybe memoize the value? Or at least put a '// BEWARE: O(N)!`?

Reply:
Done.